File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
from sys import stdin
from math import ceil, log2
input = stdin.readline

def init(node,st,ed):
    if st == ed:
        tree[node] = num[st]
        return tree[node]
    mid = (st + ed) // 2
    tree[node] = init(2*node,st,mid) + init(2*node+1,mid+1,ed)
    return tree[node]

def segment_sum(node,st,ed,l,r):
    if l > ed or r < st:
        return 0
    if l <= st and ed <= r:
        return tree[node]
    mid = (st + ed) // 2
    return segment_sum(node*2,st,mid,l,r) + segment_sum(node*2+1,mid+1,ed,l,r)

def update(node,st,ed,idx,diff):
    if idx < st or idx > ed:
        return
    tree[node] += diff
    if st != ed:
        mid = (st + ed) // 2
        update(node*2,st,mid,idx,diff)
        update(node*2+1,mid+1,ed,idx,diff)

n,m,k = map(int,input().split())

h = ceil(log2(n))
tree_size = 1 << (h+1)
tree = [0] * tree_size

num = []
for _ in range(n):
    num.append(int(input()))
    
init(1,0,n-1)
    
for _ in range(m+k):
    a,b,c = map(int,input().split())
    if a == 1:
        diff = c - num[b-1]
        num[b-1] = c
        update(1,0,n-1,b-1,diff)
    elif a == 2:
        print(segment_sum(1,0,n-1,b-1,c-1))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.util.*;
import java.io.*;

public class Main {
    static long[] num, tree;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        M += K;
        num = new long[N];
        for (int i = 0; i < N; i++) {
            num[i] = Long.parseLong(br.readLine());
        }
        int h = (int) Math.ceil(Math.log(N) / Math.log(2));
        int tree_size = (1 << (h + 1));
        tree = new long[tree_size];
        init(1, 0, N - 1);
        StringBuilder sb = new StringBuilder();
        while (M-- > 0) {
            st = new StringTokenizer(br.readLine());
            int cmd = Integer.parseInt(st.nextToken());
            if (cmd == 1) {
                int index = Integer.parseInt(st.nextToken());
                long val = Long.parseLong(st.nextToken());
                update(1, 0, N - 1, index - 1, val);
            } else {
                int left = Integer.parseInt(st.nextToken());
                int right = Integer.parseInt(st.nextToken());
                sb.append(query(1, 0, N - 1, left - 1, right - 1)).append("\n");
            }
        }
        System.out.println(sb);
    }

    static void init(int node, int start, int end) {
        if (start == end) {
            tree[node] = num[start];
        } else {
            init(node * 2, start, (start + end) / 2);
            init(node * 2 + 1, (start + end) / 2 + 1, end);
            tree[node] = tree[node * 2] + tree[node * 2 + 1];
        }
    }

    static long query(int node, int start, int end, int left, int right) {
        if (left > end || right < start) {
            return 0;
        }
        if (left <= start && end <= right) {
            return tree[node];
        }
        long leftSum = query(node * 2, start, (start + end) / 2, left, right);
        long rightSum = query(node * 2 + 1, (start + end) / 2 + 1, end, left, right);
        return leftSum + rightSum;
    }

    static void update(int node, int start, int end, int index, long val) {
        if (index < start || index > end) {
            return;
        }
        if (start == end) {
            num[index] = val;
            tree[node] = val;
            return;
        }
        update(node * 2, start, (start + end) / 2, index, val);
        update(node * 2 + 1, (start + end) / 2 + 1, end, index, val);
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }
}